File: packages/shared/src/runtime-flags.ts
import { getDb } from "./db.js";
import { getLogger, type Logger } from "./logger.js";

// Lazy-init: see llm-usage.ts. Avoids eager env validation on import.
let _logger: Logger | undefined;
function log(): Logger {
  if (!_logger) _logger = getLogger("shared");
  return _logger;
}

export type FlagValue = boolean | number | string;

/**
 * Read a runtime flag from the `runtime_flags` table, falling back to the
 * provided default when the row is missing or unreadable. Cached for the
 * lifetime of the process; agents are short-lived crons so they pick up new
 * values on the next run.
 *
 * Values are stored as JSONB so a boolean comes back as `true`, a number as
 * `5.0`, a string as `"birefnet"`. We cast through `FlagValue` to keep callers
 * honest at the type boundary.
 */
const cache = new Map<string, FlagValue>();

export async function getRuntimeFlag<T extends FlagValue>(
  key: string,
  fallback: T,
): Promise<T> {
  if (cache.has(key)) return cache.get(key) as T;
  try {
    const db = getDb();
    const { data, error } = await db
      .from("runtime_flags")
      .select("value")
      .eq("key", key)
      .maybeSingle();
    if (error) {
      log().warn({ err: error, key }, "runtime_flag read failed; using fallback");
      cache.set(key, fallback);
      return fallback;
    }
    if (!data) {
      cache.set(key, fallback);
      return fallback;
    }
    const value = data.value as T;
    cache.set(key, value);
    return value;
  } catch (err) {
    log().warn({ err, key }, "runtime_flag read threw; using fallback");
    cache.set(key, fallback);
    return fallback;
  }
}

/** Test-only — clears the in-process cache. */
export function _resetRuntimeFlagCache(): void {
  cache.clear();
}

File: packages/shared_py/runtime_flags.py
"""Dashboard-controlled runtime flags backed by the `runtime_flags` table.

Each agent reads flag values at the start of a run via `get_runtime_flag`,
falling back to the provided default if the row is missing or the read fails.
Values are cached per process; agents are short-lived crons so they pick up
new dashboard toggles on the next scheduled run.
"""

from __future__ import annotations

from .db import get_db
from .logger import get_logger

_log = get_logger("shared")
_cache: dict[str, object] = {}


def get_runtime_flag[T: (bool, int, float, str)](key: str, fallback: T) -> T:
    if key in _cache:
        return _cache[key]  # type: ignore[return-value]
    try:
        resp = (
            get_db().table("runtime_flags").select("value").eq("key", key).maybe_single().execute()
        )
        if resp is None or not getattr(resp, "data", None):
            _cache[key] = fallback
            return fallback
        value = resp.data["value"]  # JSONB → native python type
        _cache[key] = value
        return value  # type: ignore[return-value]
    except Exception as exc:  # noqa: BLE001
        _log.warning("runtime_flag_read_failed", key=key, error=str(exc))
        _cache[key] = fallback
        return fallback


def _reset_runtime_flag_cache() -> None:
    """Test-only — clears the in-process cache."""
    _cache.clear()
